Storage/src/Connection/Rest.php

```diff
@@ -600,4 +636,201 @@ protected function supportsBuiltinCrc32c()
+    {
+        $requestHash = Uuid::uuid4()->toString();
+        $args['restOnRetryExceptionFunction'] = function (
+            \Exception $e,
+            $currentAttempt,
+            &$arguments
+        ) use ($requestHash) {
+            // Since we the the last attempt number here, so incrementing it
+            // to get the current attempt count.
+            // We're adding a '2' and not '1' as we need to incorporate the initial
+            // request as well.
+            $this->updateRetryHeaders(
+                $arguments,
+                $requestHash,
+                $currentAttempt + 2
+            );
+        };
+
+        $args['restOnExecutionStartFunction'] = function (&$arguments) use ($requestHash) {
+            $this->updateRetryHeaders(
+                $arguments,
+                $requestHash
+            );
+        };
+        return $args;
+    }
+
+    /**
+     * Updates the api client identification header value with UUID
+     * and retry count
+     *
+     * @param array &$arguments The arguments array(passed by reference) used by
+     * execute method of ExponentialBackoff object.
+     * @param string $requestHash A UUID4 string value that represents a request and
+     * it's retries.
+     * @param int $currentAttempt The original attempt is of a value 1, and retries are
+     * 2, 3 and so on.
+     * @return void
+     */
+    private function updateRetryHeaders(
+        &$arguments,
+        $requestHash,
+        $currentAttempt = 1
+    ) {
+        $valueToAdd = sprintf("gccl-invocation-id/%s", $requestHash);
+        $this->updateHeader(
+            RequestWrapper::HEADER_API_CLIENT_IDENTIFICATION,
+            $arguments,
+            $valueToAdd
+        );
+
+        $valueToAdd = sprintf("gccl-attempt-count/%s", $currentAttempt);
+        $this->updateHeader(
+            RequestWrapper::HEADER_API_CLIENT_IDENTIFICATION,
+            $arguments,
+            $valueToAdd,
+            false
+        );
+    }
+
+    /**
+     * Amends the given header key with new value for a request such that
+     * the $request headers aren't modified directly and instead $options array
+     * which are applied to the request just before sending it at core level.
+     * Thus the $request object remains the same between each retry request at
+     * RequestWrappers' level.
+     *
+     * @param string $headerLine The header line to update.
+     * @param array &$arguments The arguments array(passed by reference) used by
+     *        execute method of ExponentialBackoff object.
+     * @param string $value The value to be ammended in the header line.
+     * @param bool $getHeaderFromRequest [optional] A flag which determines if
+     *        existing header value is read from $request or from $options. It's
+     *        useful to read from $options incase we update multiple values to a
+     *        single header key.
+     */
+    private function updateHeader(
```

Comment:
This function also seems generic to retrying logic. Let's see about where we can put this to make it generally available for other components as well.

Reply:
Tried in the PR mentioned above(https://github.com/googleapis/google-cloud-php/pull/6030)